[쇼핑 주문 3, 4단계] 이든 미션 제출합니다.

devfeijoa opened the pull request:
# 셀프 체크리스트

안녕하세요, 말리빈 ! 이든입니다.
늘 귀중한 시간내어 리뷰해주셔서 감사합니다 ! 

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [ ] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

# 어떤 부분에 집중하여 리뷰해야 할까요?

DM으로 말했다싶이 현재 3단계까지만 구현되어있는 상태에서 몸이 아파 이 상태에서 리뷰요청을 보내는 점 죄송하고 양해 부탁드립니다 😢
빠진 부분 빠르게 적용하여 다시 리뷰 요청 드리겠습니다 ! 🙇🏻‍♀️

 <!--
 학습 목표에 맞게 구현하였는지 점검하고, 피드백을 받을 준비가 되었는지 확인하세요.
 정답을 묻기보다, 자신의 생각과 의사 결정 과정을 공유하며 피드백을 통해 논의하는 것이 좋습니다.
 -->
 
-

## 코드 리뷰 커뮤니케이션

 <!--
 리뷰어에게 질문이 있다면, PR에 직접 코멘트를 남겨 보세요.  
 특정 코드 줄에 의견을 남기면 보다 구체적인 피드백을 받을 수 있습니다.  
 -->

📌 [GitHub에서 PR에 댓글 남기는 방법](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request)

## 참고 자료

 <!-- 관련 문서나 참고할 만한 자료를 추가하세요. -->

## 스크린샷

 <!-- 필요하다면 스크린샷을 첨부하세요. -->

![]()

devfeijoa committed:
remove: 사용하지 않는 파일과 함수 제거

devfeijoa committed:
fix: 상세 페이지에서 수량 변경 시 상품 목록에서 추가,제거가 되지 않는 오류 개선

devfeijoa committed:
remove: 사용하지 않는 파일 제거

devfeijoa committed:
fix: 장바구니 아이템이 삭제되지 않는 문제 수정

devfeijoa committed:
refactor: ListAdaptor, DiffUtil로 개선

devfeijoa committed:
feat: 장바구니 상품 전체 체크 기능 구현

- 장바구니 목록에 CartGoodsItem을 사용하도록 CartAdapter, CartViewModel, CartViewHolder, CartDiffUtil 수정
- CartViewModel에 전체 상품 선택/해제 기능(changeAllChecked) 추가

devfeijoa committed:
fix: 장바구니 상품 수량 변경 시 UI 업데이트 안되는 버그 수정

devfeijoa committed:
docs: step3 기능 명세서 업데이트

devfeijoa committed:
refactor: ViewModelFactory 구현 방식 변경

- 기존의 ViewModelFactory 유틸 클래스를 제거하고 각 ViewModel에 맞는 Factory 클래스를 생성하여 의존성 주입 방식을 변경
- ShoppingApplication 클래스에서 ViewModelFactory 인스턴스를 직접 생성하던 로직을 제거하고 각 Activity에서 Factory를 통해 ViewModel을 생성하도록 수정

devfeijoa committed:
refactor: Retrofit 서비스 객체 접근 방식 변경

기존의 `getRetrofitService()` 메서드를 통해 서비스 객체에 접근하던 방식을 제거하고 대신 `service` 프로퍼티를 `public`으로 변경하여 직접 접근하도록 수정

devfeijoa committed:
refactor: ViewModel에서 콜백 방식 제거하고 코루틴으로 변경

devfeijoa committed:
feat: 상품 상세 정보 응답 DTO 구현

devfeijoa committed:
refactor: ProductService의 API 요청 함수를 코루틴으로 변경하고, ProductRepository에 Result 패턴 적용

devfeijoa committed:
refactor: 코루틴 적용하여 비동기 처리 개선

devfeijoa committed:
refactor: 코루틴 적용 및 로직 수정

- 상품 상세 정보 요청 시 suspend 함수로 변경
- 장바구니 관련 비동기 처리 수정
- 최근 본 상품 로직 수정 및 코루틴 적용
- 상품 상세 화면 UI 바인딩 오류 수정
- 상품 상세 화면에서 장바구니 담기 성공 시 장바구니 화면으로 이동

devfeijoa committed:
refactor: 상품 수량 기본값 1로 변경

devfeijoa committed:
fix: 상품 상세 화면에서 최근 본 상품 표시 로직 수정 및 장바구니 이동 오류 수정

- 최근 본 상품 정보를 불러온 후 비동기 문제 해결하여 가시성 업데이트 하도록 순서 변경
- 최근 본 상품 정보가 없을 경우 UI 업데이트하지 않도록 수정
- 장바구니 담기 성공 시 장바구니 화면으로 이동하도록 수정

malibinYun commented on app/src/main/java/woowacourse/shopping/data/local/history/HistoryDao.kt:
```diff
@@ -14,7 +14,7 @@ interface HistoryDao {
 
     @Query("SELECT * FROM history WHERE id = :id LIMIT 1")
-    fun findById(id: Int): HistoryEntity?
+    fun findById(id: Int): HistoryEntity
```
반환 값이 non-null으로 바뀐 이유는 무엇인가요?
History가 0개인 경우는 반드시 없을까요?

devfeijoa replied:
처음에 값이 안들어갔을 때는 History가 0개일 것 같습니다 !  해당 피드백 적용하여 수정했습니다 !

57f43d6a9a77cf9f1cbec42b33bf4b0850cc38c1

malibinYun commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -54,77 +65,73 @@ class GoodsViewModel(
-        historyRepository.getAll { histories ->
+        viewModelScope.launch(Dispatchers.IO) {
+            val histories = historyRepository.getAll()
+
```
Dispatchers.IO를 해줘야 에러가 발생하지 않으셨겠지요?
왜그럴까요?
suspend 를 붙였는데도 왜 livedata의 setValue를 호출하면 문제가 생기는걸까요?
스레드 이동에 대해 충분히 고민해보세요.
또. dao에 suspend를 붙인 것과 붙이지 않은 것은 어떤 차이인지 꼭 공부해보세요 :)

그리고 라이브러리들을 활용해서 setValue를 사용하면 문제가 안되는데, dao에 suspend를 붙이지 않은 것과, 이전 callback으로 구성했을 때 setValue를 하면 문제가 생기는 이유에 대해서 정확하게 짚고 넘어가시면 좋겠어요.

devfeijoa replied:
1. Dispatchers.IO를 해줘야 에러가 발생하지 않았던 이유는
DB 접근(DAO)과 같은 I/O 작업은 Main Thread에서 할 수 없으므로(안드로이드 UI는 오직 Main Thread에서만 수정할 수 있어서 시간이 오래걸리는 작업인(네트워크 통신, DB접근) 을 메인 스레드에서 수행하면 ANR이 발생하기 때문), Dispatchers.IO를 통해 코루틴이 실행될 스레드를 백그라운드로 전환해줘야 에러가 발생하지 않기 때문입니다 !



2. suspend를 붙였는데도 왜 livedata의 setValue를 호출하면 문제가 생기는 이유는,

suspend 함수는 스레드를 바꾸는게 아니라 "오래 걸릴 수 있으니 실행을 잠시 멈추고 스레드를 차단(Block)하지 않은 채 다른 일을 하다가 결과가 준비되면 다시 돌아와서 이어서 실행할 수 있다"는 뜻입니다.

따라서 suspend는 코루틴을 '중단 가능'하게 만들 뿐, 어떤 스레드에서 실행될지는 코루틴을 시작한 CoroutineContext (주로 
Dispatcher)가 결정합니다.

스레드 관리는 Dispatcher가 하며 현재 어느 스레드에 있는지 항상 인지해야 합니다 ! 따라서 setValue는 Main Thread에서만 호출해야 합니다 !




3. DAO에 suspend를 붙인것과 붙이지 않은것의 차이는 다음과 같습니다..!

3-1. suspend를 붙인 경우
Room은 이 함수가 suspend인 것을 보고 내부적으로 알아서 I/O 스레드에서 쿼리를 실행해 줍니다.
Room이 스레드 전환을 해주기 때문에 Main Thread를 블로킹하지 않습니다 !

3-2. suspend가 없는 경우
일반적인 함수입니다
호출 즉시 결과를 반환할 때까지 스레드를 멈춥니다(블로킹합니다)
만약 이 함수를 Main Thread에서 호출하게 된다면NetworkOnMainThreadException이 발생하게 됩니다.
따라서 이렇게 사용하려면 직접 withContext(Dispatchers.IO)로 감싸서 백그라운드 스레드로 보내야만 합니다 !

malibinYun commented on app/src/main/java/woowacourse/shopping/data/remote/cart/CartRepository.kt:
```diff
@@ -1,205 +1,76 @@
+                response.body()
+            } else {
+                throw HttpException(response)
+            }
```
throw를 통해서 result 객체를 만들기보다, 내가 직접 에러 제어를 수행한 뒤, Result 객체를 생성해서 반환해보는 건 어떨까요?

devfeijoa replied:
제가 이 피드백을 최대한 이해한 바로는, 
if/else: 예측 가능한 실패 (HTTP 에러)를 값으로 처리
catch: 예측하지 못한 예외 (네트워크 단절, DNS 오류 등)를 처리하라는 것으로 이해하였습니다..!

예측 가능한 실패 (400,500번대) 는 실패했다고 보고하는 Result.failure를 사용하고, 예측하지 못하는 예외 (네트워크 단절)은 try-catch로 잡으라는 것으로 이해하였습니다..!

if-else로 성공인지 실패인지 상태를 확인하고 , 실패의 경우에는 각 경우에 맞는 Result 객체를 직접 만들어 반환하라는 의도였을까요..?

malibinYun commented on app/src/main/java/woowacourse/shopping/data/remote/cart/CartService.kt:
```diff
@@ -12,33 +12,33 @@ import retrofit2.http.Query
         @Query("page") page: Int = 0,
         @Query("size") size: Int = 5,
-    ): Call<CartResponse>
+    ): Response<CartResponse>
```
CartResponse를 바로 반환하지 않고 Response 객체로 한 번 더 감싼 이유는 무엇인가요?

devfeijoa replied:
해당 피드백 아래에 답변하였습니다 !

malibinYun commented on app/src/main/java/woowacourse/shopping/feature/cart/CartViewModel.kt:
```diff
@@ -66,14 +77,22 @@ class CartViewModel(
         }
     }
 
+    fun changeAllChecked() {
```
메서드명만 보고 어떤 일을 수행하고자 하는 것인지 다른 개발자들도 한 번에 알 수 있을까요?

devfeijoa replied:
해당`changeAllChecked()` 메서드명을 `toggleAllItemsChecked()`로 변경하여 더 명확한 메서드명으로 수정하였습니다 ! 

df09cbe2a7e888460d62e17b890c95e17782c62b

malibinYun commented on app/src/main/java/woowacourse/shopping/feature/cart/CartViewModel.kt:
```diff
@@ -93,143 +112,159 @@ class CartViewModel(
+                    Log.e("addCartTest", "장바구니 추가 실패")
                 }
         }
     }
```
인덴트가 매우 많이 발생하네요.
레벨 1에서 배웠던 객체지향 생활체조를 잊지 마시고 적용해보세요 :)

devfeijoa replied:
해당 피드백 적용하여 수정하였습니다 ! 

2e65bca23ee272ffebe593228e8c244d11527491

malibinYun commented on app/src/main/java/woowacourse/shopping/feature/cart/CartViewModel.kt:
```diff
@@ -93,143 +112,159 @@ class CartViewModel(
-                                    .filter {
+                                response
+                                    ?.content
+                                    ?.filter {
                                         it.category == category &&
-                                            it.id.toInt() != latestProduct.product.id &&
-                                            it.id !in cartProductIds
-                                    }.take(10)
+                                            it.id.toInt() != latestProduct?.product?.id &&
+                                            cartProductIds?.contains(it.id) != true
+                                    }?.take(10)
 
-                            _recommendItems.value =
-                                recommendProducts.map {
-                                    Cart(id = 0, product = it.toDomain(), quantity = 0)
-                                }
-                        },
-                        onError = {
+                            withContext(Dispatchers.Main) {
+                                _recommendItems.value =
+                                    recommendProducts?.map {
+                                        Cart(id = 0, product = it.toDomain(), quantity = 0)
+                                    }
+                            }
+                        }.onFailure {
                             Log.e("loadProductsInRange", "상품 요청 실패", it)
-                        },
-                    )
-                },
-                onError = {
+                        }
+                }.onFailure {
```
코루틴을 사용하면 콜백 지옥을 벗어난다고 하던데, 벗어나고 있지 못하고 있네요.
어떻게 개선할 수 있을지 고민해보시면 좋겠네요 :)

devfeijoa replied:
해당 피드백 적용하여 수정하였습니다 ! 

2e65bca23ee272ffebe593228e8c244d11527491

malibinYun commented on app/src/main/java/woowacourse/shopping/feature/goods/GoodsViewModel.kt:
```diff
@@ -176,51 +184,55 @@ class GoodsViewModel(
 
     private fun loadHistories() {
-        historyRepository.getAll { allHistories ->
-            histories.postValue(allHistories)
+        viewModelScope.launch(Dispatchers.IO) {
+            val allHistories = historyRepository.getAll()
+            withContext(Dispatchers.Main) {
+                histories.value = allHistories
+            }
             refreshItems()
```
dispatcher의 변화를 매번 viewmodel에서 수행해주어야할까요?
프로젝트가 커지다보면 같은 데이터를 가져오는 코드는 점점 많아지겠지요.

devfeijoa replied:
매번 viewModel에서 dispatcher를 지정해주다보면 위의 코드처럼 dispatcher를 매번 변경해주어야 하는 일이 발생합니다..!

때문에 Dispatcher를 지정해주어야 한다면, API를 호출하기 시작하는 시점인 Reppository에서 Dispatcher를 지정해주는 것이 맞을 것 같습니다..!

다만 ROOM 또한 레트로핏처럼 내부적으로 백그라운드 스레드로 동작하도록 스위칭 해주는 것 같아 디스패처를 변경하지 않았습니다 !

따라서 해당 피드백 반영하여 수정하였습니다 ! 

081734a4c5a2c581d7dffbeffa341c3d61076b63

malibinYun reviewed (changes requested):
3단계 미션 고생 많으셨어요.
여러 생각해볼 점들에 코멘트 남겨두었으니 확인 부탁드려요.
다음 미션에는 꼭 4단계까지 모두 반영하고 제출 해보세요!

devfeijoa committed:
feat: item_coupon.xml 파일 구현

devfeijoa committed:
feat: 쿠폰 관련 결제 화면 문자열 리소스 추가

devfeijoa committed:
refactor: 상품 목록 화면 구분선 추가

devfeijoa committed:
feat: 상품 목록 화면 RecyclerView 아이템 타입 정의

- GoodsRvItems: RecyclerView 아이템의 sealed class
  - HistoryItem: 최근 본 상품 목록
  - DividerItem: 구분선
  - GoodsItem: 일반 상품
  - LoadMoreItem: 더 보기 버튼

devfeijoa committed:
feat: GoodsDiffUtil 추가

devfeijoa committed:
feat: GoodsProduct UI 모델 추가

devfeijoa committed:
refactor: 상품 목록 화면 버그 수정

devfeijoa committed:
feat: 최종 주문 페이지 추가

devfeijoa committed:
feat: 쿠폰 조회 API 응답 DTO 추가

- CouponResponse
- CouponResponseItem
- AvailableTime

devfeijoa committed:
feat: 주문 API 서비스 및 클라이언트 구현

devfeijoa committed:
feat: 주문 화면 및 쿠폰 도메인 모델 구현

- OrderActivity 및 관련 레이아웃 추가
- Coupon, BogoCoupon, FixedCoupon, FreeshippingCoupon, MiracleSaleCoupon 도메인 모델 추가
- Payment, AvailableTime 도메인 모델 추가
- OrderViewModel, OrderUistate, CouponState 구현
- Coupon 관련 API 응답 DTO (BogoCouponResponse, FixedCouponResponse, FreeshippingCouponResponse, MiracleSaleCouponResponse, CouponsResponse) 추가
- CouponService, CouponRepository, OrderRepository 추가
- 장바구니 화면에서 주문하기 버튼 클릭 시 OrderActivity로 이동하는 기능 추가
- CartViewModel의 carts LiveData 타입을 Carts로 변경하여 총 수량 정보 포함하도록 수정

refactor: CouponResponseItem 제거

- 현재 사용되지 않는 것으로 판단되어 제거

refactor: ProductResponse에 toDomain() 확장 함수 추가

refactor: HistoryRepositoryImpl 불필요한 import 제거

refactor: DividerViewHolder import 정리

malibinYun commented on app/src/main/java/woowacourse/shopping/data/local/ShoppingDatabase.kt:
```diff
@@ -26,7 +26,8 @@ abstract class ShoppingDatabase : RoomDatabase() {
                         context.applicationContext,
                         ShoppingDatabase::class.java,
                         "app_database",
-                    ).build()
+                    ).fallbackToDestructiveMigration(true)
```
fallbackToDestructiveMigration를 활용한 이유는 무엇인가요?
어떤 역할을 수행하나요? 그로인해 발생할 수 있는 사이드이펙트는 무엇이 있을까요?

devfeijoa replied:
데이터베이스 구조가 바뀌었는데, 어떻게 바꿀지(Migration) 알려주지 않으면 그냥 기존 데이터베이스를 전부 날리고 새로 만들어라는 것입니다 ! 

하지만 가장 치명적인 사이드 이펙트는 "사용자 데이터의 완전한 소실" 이 있을 수 있어 주의해서 사용해야 할 것 같습니다..!

해당  함수의 사이드 이펙트에 대해서는 크게 생각해보지 않았는데, 말리빈 덕분에 좀 더 주의해서 써야함을 알게되었습니다 ! 감사합니다 👍

malibinYun commented on app/src/main/java/woowacourse/shopping/data/local/history/repository/HistoryRepositoryImpl.kt:
```diff
@@ -1,35 +1,26 @@
-                dao.deleteOldest()
-            }
-            dao.insert(cart.toHistoryEntity())
+    override suspend fun insert(product: Product) {
+        if (dao.getAll().size >= 10) {
+            dao.deleteOldest()
         }
+        dao.insert(product.toHistoryEntity())
     }
```
10개 제한을 걸어두는 것은 어느 레이어의 역할일까요?
다른 관점으로는, 저장할 때 꼭 10개 제한을 걸어두어야할까요?
꺼내올 때만 10개 제한을 걸어두는 방법은 어떻게 생각하시나요?

malibinYun commented on app/src/main/java/woowacourse/shopping/data/remote/cart/CartRepository.kt:
```diff
@@ -1,205 +1,79 @@
+    private val cartService: CartService,
+) {
+    suspend fun fetchAllCart(): Result<CartResponse?> =
+        runCatching {
+            val response = cartService.requestCart(page = null, size = null)
+            if (response.isSuccessful) {
+                response.body()
+            } else {
+                throw HttpException(response)
+            }
+        }
```
https://github.com/woowacourse/android-shopping-order/pull/137#discussion_r2147408818
> 제가 이 피드백을 최대한 이해한 바로는,
if/else: 예측 가능한 실패 (HTTP 에러)를 값으로 처리
catch: 예측하지 못한 예외 (네트워크 단절, DNS 오류 등)를 처리하라는 것으로 이해하였습니다..!
>
> 예측 가능한 실패 (400,500번대) 는 실패했다고 보고하는 Result.failure를 사용하고, 예측하지 못하는 예외 (네트워크 단절)은 try-catch로 잡으라는 것으로 이해하였습니다..!
>
> if-else로 성공인지 실패인지 상태를 확인하고 , 실패의 경우에는 각 경우에 맞는 Result 객체를 직접 만들어 반환하라는 의도였을까요..?

runCatching이 반환 값을 어떻게 만들어내는지를 잘 들여다보시면 좋겠어요.
그러면 내부에서 throw를 하더라도 앱이 크래시가 나는 것이 아닌, Result 객체를 반환하는 원리를 알게되겠지요.
그 이후에 다시 이전 코멘트를 보시면 이해하실 수 있을 거란 생각이 드네요 :)

devfeijoa replied:
runCatching 내부에서 아래와 같이  try-catch를 사용하고있는 것을 확인했습니다.
```
public inline fun <T, R> T.runCatching(block: T.() -> R): Result<R> {
    return try {
        Result.success(block())
    } catch (e: Throwable) {
        Result.failure(e)
    }
}
```
인자로 전달받은 함수를 수행한 후 그 결과에 따라 Result를 생성해 화반환합니다. 이 과정에서 에러가발 생할 경우 캐치 구문에서 Failure 에 에러를 담아 반환합니다. 

>throw를 통해서 result 객체를 만들기보다, 내가 직접 에러 제어를 수행한 뒤, Result 객체를 생성해서 반환해보는 건 어떨까요?

말리빈께서 말씀하신 것을 고민해본 결과 다음과 같이  try-catch를 사용해 직접 Result를 반환하는 것이 라고 생각했는데 이 경우 기존에 runCatching을 사용하는 것과 차이가 없다고 생각하는데 혹시 다른 의미가있으신지 의견을 부탁드려도 될까요 ? 😢 

```
suspend fun fetchAllCart(): Result<CartResponse?> =
        try {
            val response = cartService.requestCart(page = null, size = null)
            Result.success(response)

        } catch (e: Exception) {
            Result.failure(e)
        }
```

928c24b055cd0c80a16a2810545a1f19b3f77851

malibinYun replied:
try-catch로 전부 바꾸라는 내용이 아닌,
runcatching을 사용하더라도 throw를 통해 Result를 생성하지 않고, 직접 Result를 만들어 반환해보라는 의미었어요 :)

malibinYun commented on app/src/main/java/woowacourse/shopping/data/remote/cart/CartRepository.kt:
```diff
@@ -1,205 +1,79 @@
-                ) {
-                    Log.e("CartRepository", "네트워크 실패", t)
-                    onError(t)
-                }
-            },
-        )
-    }
+            val locationHeader =
+                response
+                    .headers()["Location"]
+                    ?.split("/")
+                    ?.last()
+                    ?.toLongOrNull()
+            if (locationHeader == null) throw Exception("Location header not found")
+            locationHeader
```
runCatching 블럭이 길어지면 길어질수록 그래서 어디에서 Excpetion이 발생하는 코드인지 알기 어렵다고 생각해요.
국소적으로 예외가 발생할 수 있는 부분만 runCatching을 활용해보는 건 어떨까요?

devfeijoa replied:
런캐칭 구문에서 에러 처리가 가능한 부분은 굳이 throw로 던지지 않아도 가능할 것 같아 해당 부분을 제거 하였습니다 ! 

d0c188627c508552e38eac35ffae0f43ef5aed53

malibinYun commented on app/src/main/java/woowacourse/shopping/data/remote/cart/CartService.kt:
```diff
@@ -12,33 +12,33 @@ import retrofit2.http.Query
-        @Query("size") size: Int = 5,
-    ): Call<CartResponse>
+        @Query("page") page: Int?,
+        @Query("size") size: Int?,
+    ): Response<CartResponse>
```
https://github.com/woowacourse/android-shopping-order/pull/137#discussion_r2136726809
여전히 유효한 피드백이에요.
CartResponse를 바로 반환하지 않고 Response 객체로 한 번 더 감싼 이유는 무엇인가요?

devfeijoa replied:
무조건 붙여야 하는줄 알았습니다..! 말리빈 덕분에 해당 부분 알게 되어 피드백 적용하였습니다 !

928c24b055cd0c80a16a2810545a1f19b3f77851

malibinYun commented on app/src/main/java/woowacourse/shopping/data/remote/coupon/CouponClient.kt:
```diff
@@ -0,0 +1,40 @@
+    val service: OrderService by lazy {
+        retrofit.create(OrderService::class.java)
+    }
+    private val okHttpClient: OkHttpClient by lazy {
+        OkHttpClient
+            .Builder()
+            .addInterceptor(HeaderInterceptor(BuildConfig.USER_ID, BuildConfig.USER_PASSWORD))
+            .addInterceptor(logging)
+            .build()
```
id와 password가 런타임에 변경된다면 유연하게 대처할 수 있는 구조일까요?

devfeijoa replied:
현재는 아이디와 패스워드가 고정되어있어 말씀하신 것 처럼 유연하게 대처하지 못할 것 같습니다..!

하지만 로그인 기능을 잘 몰라서 어떻게 개선해야할 지 열심히 찾아보았지만 이해가 잘 되지않아 아직도 모르겠습니다..! 혹시 관련 힌트나 문서를 여쭈어봐도 될까요..? 😢

malibinYun replied:
레벨 1때 랜덤 테스트를 어떤식으로 가능하도록 구성했었는지 되돌아보면 좋겠네요 :)

malibinYun commented on app/src/main/java/woowacourse/shopping/data/remote/coupon/CouponRepository.kt:
```diff
@@ -0,0 +1,12 @@
+class CouponRepository(
+    val couponService: CouponService,
+) {
+    suspend fun getCoupons(): Result<List<Coupon>> =
+        runCatching {
+            couponService.getCoupons().map { it.toDomain() }
+        }
```
이 코드에서 Exception이 발생할 수 있는 경우가 있나요?
runCatching을 감싸놓은 의도가 궁금하네요.

devfeijoa replied:
1. 네트워크 통신 자체의 실패
2. JSON파싱 실패
3. 데이터 변환(toDomain) 실패
가 있을 수 있어 이런 부분들에서 예외가 발생할 수 있다고 생각하였습니다 ! 따라서 runCatching으로 감싸두었습니다. 

하지만 말리빈의 질문을 듣고 답변을 하다보니, 이 중에 어떤 실패인지 알기 어렵다는 느낌이 들었는데요..! 혹시 어떤 실패인지 구분하기 어렵다는 의도로 질문하셨던것일까요...? 그렇다면 후에 실패의 원인이 무엇인지 명확히 알 수 있도록 예외처리를 하여야 할 것 같네용..!

malibinYun replied:
> 네트워크 통신 자체의 실패

이전엔 Response\<T> 로 감싸져있었기에, 발생할 일이 없었어요. 

> JSON파싱 실패
데이터 변환(toDomain) 실패

위 두 부분에서는 어떻게 발생할 수 있었나요?
서버에서 이상한 json을 응답하는 경우를 막는 의도인가요?
데이터의 변환은 직접 짠 코드인데, 어느 부분에서 exception이 발생하나요?

정말로 발생하는 Exception인지를 먼저 알아보시면 좋겠어요.
모든게 의심된다면 모든 코드를 try-catch 안에 다 집어넣어야하니까요.

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/model/Carts.kt:
```diff
@@ -1,6 +1,21 @@
     val totalQuantity: Int,
-)
+) : Serializable {
+    val buyCarts: List<Cart>
+        get() = carts.filter { it.isChecked }.map { it.cart }
```
buyCarts를 접근할 때마다 매번 불필요한 연산을 반복할 거란 생각이 드네요.
한 번만 연산하도록 변경해보는 건 어떨까요?

devfeijoa replied:
get()을 제거하고 불필요한 연산을 반복하지 않도록 수정하였습니다 ! 

9876515e98ec38e07b95d9d736883664510245c0

malibinYun commented on app/src/main/java/woowacourse/shopping/feature/cart/CartViewModel.kt:
```diff
@@ -93,143 +114,161 @@ class CartViewModel(
                             val matchedProduct =
-                                response.content.find { it.id.toInt() == latestProduct.product.id }
+                                response?.content?.find { it.id.toInt() == latestProduct?.id }
                             val category = matchedProduct?.category
 
                             val recommendProducts =
-                                response.content
-                                    .filter {
+                                response
+                                    ?.content
+                                    ?.filter {
                                         it.category == category &&
-                                            it.id.toInt() != latestProduct.product.id &&
-                                            it.id !in cartProductIds
-                                    }.take(10)
+                                            it.id.toInt() != latestProduct?.id &&
+                                            cartProductIds?.contains(it.id) != true
+                                    }?.take(10)
 
-                            _recommendItems.value =
-                                recommendProducts.map {
-                                    Cart(id = 0, product = it.toDomain(), quantity = 0)
-                                }
-                        },
-                        onError = {
+                            withContext(Dispatchers.Main) {
+                                _recommendItems.value =
+                                    recommendProducts?.map {
+                                        Cart(id = 0, product = it.toDomain(), quantity = 0)
+                                    }
+                            }
```
코루틴을 활용했을 때 특징 중 하나로 꼽는 것은 콜백 지옥을 제거할 수 있다는 것이지요.
그러나 Result를 사용하면서 여러 repository를 사용해야할때 콜백 지옥을 피할 수 없어보이네요.
Result를 사용하지 않고 구조를 변경한다면 어떤 전략을 사용해볼 수 있을까요?
특히 에러처리를 할 때도 집중해보시면 좋겠네요.

devfeijoa replied:
해당 피드백 적용하여 수정하였습니다 ! 

2e65bca23ee272ffebe593228e8c244d11527491

malibinYun commented on app/src/main/java/woowacourse/shopping/feature/goods/adapter/GoodsRvItems.kt:
```diff
@@ -0,0 +1,20 @@
+import woowacourse.shopping.domain.model.Product
+import woowacourse.shopping.feature.goods.GoodsProduct
+
+sealed class GoodsRvItems(
+    val viewType: ItemViewType,
```
레벨1때 배웠던 축약어의 사용 지양을 늘 잊지 마세요 :)
xml의 view id도 마찬가지랍니다. 의미있는 이름을 지어보도록 연습해보세요.
RecyclerView에서 ListView로 변경해야한다면 LvItems로 이름을 바꿔야할까요?

devfeijoa replied:
축약어를 사용하지 않고 의미있는 이름으로 수정하였습니다 !
`GoodsRvItems`를 `ProductFeedItem`으로 변경하였습니다 ! 

081734a4c5a2c581d7dffbeffa341c3d61076b63

malibinYun reviewed (changes requested):
3, 4단계 미션 고생 많으셨어요.
여러 고민해볼 점들에 코멘트 남겨두었으니 깊게 생각해보는 시간 가져보세요.
피드백 반영 후 다시 리뷰 요청 부탁드려요 :)

malibinYun commented:
아직 피드백 반영이 많이 되었다는 느낌은 못받았네요.
약속대로 1번 더 요청을 받을게요.
다음 요청에 피드백 반영 등이 되어있지 않았다면 반영에 대한 의사가 없는 것으로 간주하고,
피드백의 필요성이 없다 판단해 리뷰 없이 머지하도록 하겠습니다 :)

devfeijoa committed:
refactor: 불필요한 연산을 하지 않도록 생성자에서 초기화하도록 개선

devfeijoa committed:
feat: 주문 화면 총 결제 금액 문자열 추가

devfeijoa committed:
feat: FreeshippingCoupon 적용 조건 및 할인 로직 구현

devfeijoa committed:
feat: FixedCoupon 할인 적용 기능 구현

devfeijoa committed:
feat: MiracleSaleCoupon 할인 적용 기능 구현

devfeijoa committed:
refactor: CartViewModel의 LiveData 타입
`List<CartGoodsItem>`으로 변경

- 이에 따라 관련 로직 및 레이아웃 바인딩을 수정했습니다.
`List<CartGoodsItem>`으로 변경하고, Intent에 ArrayList로 담도록 수정했습니다.

devfeijoa committed:
refactor: Cart, Product, and CartGoodsItem Serializable 구현

devfeijoa committed:
fix: OrderActivity 이동 및 장바구니 상품 전달 로직 수정

- CartRecommendFragment에서 OrderActivity로 이동할 때 `startActivity`를 호출하도록 수정
- OrderActivity의 `newIntent` 메서드에서 `CartActivity` 대신 `OrderActivity`를 참조하도록 수정

devfeijoa committed:
refactor: 배송비 텍스트 위치 변경

devfeijoa committed:
feat: TextView 바인딩 어댑터 추가

- `expirationDate`: LocalDate를 특정 형식의 문자열로 변환하여 표시
- `priceText`: 가격을 통화 형식으로 변환하여 표시
- `minimumOrderPriceText`: 쿠폰 종류에 따라 최소 주문 금액을 표시하고, 해당하지 않는 경우 뷰를 숨김

devfeijoa committed:
feat: 만료일 표시를 위한 문자열 리소스 추가

devfeijoa committed:
refactor: 쿠폰 모델의 order 파라미터 타입 변경

- `Carts`에서 `List<Cart>`로 변경
- 이에 따라 `BogoCoupon`의 `findTargetProductForBogo` 메서드를 `BogoCoupon` 내부로 이동

devfeijoa committed:
feat: CouponAdapter와 관련 로직 구현

- 쿠폰 목록을 표시하기 위한 RecyclerView 어댑터를 추가
- `CouponViewHolder`, `CouponClickListener`, `CouponDiffUtil`을 함께 구현하여 쿠폰 아이템의 클릭 이벤트 처리 및 UI 업데이트를 관리합니다.
- `CouponClient`의 서비스 타입을 `CouponService`로 변경

devfeijoa committed:
feat: 주문 기능 구현

- 주문 화면 레이아웃 및 데이터 바인딩 설정
- OrderViewModel에서 쿠폰 정보를 가져오고, 주문 UI 상태를 관리
- OrderRepository를 통해 주문 요청 API 호출
- 쿠폰 선택 기능 및 관련 UI 업데이트 로직 추가
- OrderViewModelFactory 구현

devfeijoa committed:
refactor: 불필요한 throw문 제거

devfeijoa committed:
refactor: API 응답 처리 방식 변경

- Retrofit `Response` 객체를 직접 반환하지 않도록 수정
- `runCatching` 대신 `try-catch`를 사용하여 예외 처리
- API 호출 성공 시 `Result.success` 반환, 실패 시 `Result.failure` 반환

devfeijoa committed:
refactor: HistoryDao.findById 반환 타입 nullable로 변경

- 해당 id를 가진 데이터가 없을 경우 null을 반환하도록 수정

devfeijoa committed:
refactor: HistoryDao의 모든 메서드에 suspend 키워드 추가

devfeijoa committed:
refactor: CartViewModel 코드 개선 및 리팩토링

- 장바구니 수량 변경 및 삭제 로직을 별도 함수로 분리
- 선택된 상품 정보 업데이트 로직 개선
- 페이지 번호 조정 로직 추가
- 코루틴 디스패처 명시 제거 (viewModelScope 기본값 사용)
- 전체 장바구니 및 상품 목록 조회 시 에러 처리 개선
- 추천 상품 생성 로직 분리 및 명확화

devfeijoa committed:
refactor: HistoryRepository.getAll() 호출 시 withContext(Dispatchers.IO) 제거

devfeijoa committed:
refactor: `GoodsRvItems`를 `ProductFeedItem`으로 이름 변경

devfeijoa committed:
refactor: 전체 선택 기능 관련 메서드명 변경

- `changeAllChecked()` 메서드명을 `toggleAllItemsChecked()`로 변경

devfeijoa commented:
안녕하세요 말리빈!

기회를 주신만큼 열심히 고민하고 피드백 반영하여 4단계까지 완성하여 구현해보았습니다..! 🙇🏻‍♀️

말리빈의 리뷰가 저에게는 많이 어려웠어서 그만큼 많이 고민하게 되었던것 같아요...! 그 과정에서 제가 얼마나 깊이 고민하지 않았었는지 반성하는 계기가 되었습니다 🥹
앞으로는 의식적으로 깊이 고민해보고, 앞으로도 더 명확한 의도와 목적을 갖고 코드를 작성해보려 합니다 ☺️

다시 한 번 감사합니다 😊

malibinYun commented on app/src/main/java/woowacourse/shopping/data/local/Mapper.kt:
필요 없어진 코드는 주석 처리 보다 삭제하는 건 어떨까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/model/BogoCoupon.kt:
```diff
@@ -0,0 +1,53 @@
+        )
+    }
+
+    fun findTargetProductForBogo(
+        standardQuantity: Int,
+        buyCarts: List<Cart>,
+    ): Int =
```
쿠폰이 직접 상품을 찾는 일을 하고 있네요. 적절한 역할 위임이었을까요?
다른 개발자들은 이 역할이 쿠폰 내에 있음을 유추할 수 있을까요?

malibinYun commented on app/src/main/java/woowacourse/shopping/domain/model/FixedCoupon.kt:
```diff
@@ -0,0 +1,38 @@
+    val minimumAmount: Int,
+    val discount: Int,
+) : Coupon {
+    override fun isUsable(
+        today: LocalDateTime,
+        order: List<Cart>,
+        payment: Int,
+    ): Boolean {
+        if (isExpired(today.toLocalDate())) return false
+
+        if (payment < minimumAmount) return false
+
+        return true
+    }
+
+    override fun applyToPayment(
+        origin: Payment,
+        order: List<Cart>,
+    ): Payment {
+        val newTotalPayment = (origin.originPayment - discount) + origin.deliveryFee
+
+        return origin.copy(
+            couponDiscount = -discount,
+            totalPayment = newTotalPayment.coerceAtLeast(0),
```
이 쿠폰에서 사용하지 않는 List\<Cart> 를 파라미터로 받고 있네요.
정말로 interface로 공통화할 수 있는 요소였나요?

malibinYun commented on app/src/main/java/woowacourse/shopping/feature/cart/CartProductFragment.kt:
```diff
@@ -17,15 +18,23 @@ import woowacourse.shopping.feature.model.ResultCode
 
 class CartProductFragment : Fragment() {
     private lateinit var binding: FragmentCartProductBinding
-    private val viewModel: CartViewModel by activityViewModels<CartViewModel>()
+    private val viewModel: CartViewModel by activityViewModels {
+        val app = (requireActivity().application as ShoppingApplication)
+        CartViewModelFactory(
+            app.cartRepository,
+            app.productRepository,
+            app.historyRepository,
+        )
+    }
     private lateinit var adapter: CartAdapter
 
     override fun onCreateView(
         inflater: LayoutInflater,
         container: ViewGroup?,
         savedInstanceState: Bundle?,
     ): View {
-        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_cart_product, container, false)
+        binding =
+            DataBindingUtil.inflate(inflater, R.layout.fragment_cart_product, container, false)
```
이 프래그먼트는 메모리릭이 발생할 수 있겠어요.
무엇이 문제일까요?
다른 프래그먼트들도 같은 문제를 가지고 있는 것들이 눈에 보이네요.

malibinYun commented on app/src/main/java/woowacourse/shopping/feature/cart/CartViewModel.kt:
```diff
@@ -85,154 +114,162 @@ class CartViewModel(
+        }
+    }
+
+    fun loadCarts() {
+        viewModelScope.launch {
+            _isLoading.value = true
+            cartRepository
+                .fetchCart(currentPage, PAGE_SIZE)
+                .onSuccess { response ->
+                    response?.let {
+                        val cartList =
+                            it.content.map { cartItem ->
+                                CartGoodsItem(
+                                    Cart(
+                                        id = cartItem.id,
+                                        product = cartItem.product.toDomain(),
+                                        quantity = cartItem.quantity,
+                                    ),
+                                )
+                            }
+                        _carts.value = cartList
+                        _page.value = it.number
+                        updatePageButtonStates(it.first, it.last, it.totalElements.toInt())
+                    }
+                }.onFailure { e ->
+                    Log.e("CartViewModel", "장바구니 로딩 실패", e)
+                }
+            _isLoading.value = false
+        }
     }
```
레벨1때 항상 강조했던 객체지향 생활체조를 잊지 않고 꼭 적용해보세요.

malibinYun reviewed (approved):
미션 완주하시느라 고생 많으셨어요 :)
여러 군데 피드백 남겨두었으니 확인해보세요!
미션은 이만 마무리하도록 하겠습니다.
지금껏 고민했던 것들이 앞으로 개발하면서 많은 도움이 되었으면 좋겠어요.
앞으로 남은 미션들과 프로젝트들도 잘 완료하시길 응원할게요 💪
